core/src/test/java/org/elasticsearch/cluster/serialization/DiffableTests.java

```diff
@@ -96,8 +97,81 @@ public TestDiffable readFrom(StreamInput in) throws IOException {
         assertThat(serialized.get("foo").value(), equalTo("1"));
         assertThat(serialized.get("baz").value(), equalTo("4"));
         assertThat(serialized.get("new").value(), equalTo("5"));
+    }
+
+    public void testImmutableOpenIntMapDiff() throws IOException {
+        ImmutableOpenIntMap.Builder<TestDiffable> builder = ImmutableOpenIntMap.builder();
+        builder.put(1, new TestDiffable("1"));
+        builder.put(2, new TestDiffable("2"));
+        builder.put(3, new TestDiffable("3"));
+        ImmutableOpenIntMap<TestDiffable> before = builder.build();
+        builder = ImmutableOpenIntMap.builder(before);
+        builder.remove(2);
+        builder.put(3, new TestDiffable("4"));
+        builder.put(4, new TestDiffable("5"));
+        ImmutableOpenIntMap<TestDiffable> after = builder.build();
+        DiffableUtils.KeySerializer<Integer> keySerializer = randomBoolean() ? DiffableUtils.getIntKeySerializer() : DiffableUtils.getVIntKeySerializer();
+        Diff diff = DiffableUtils.diff(before, after, keySerializer);
+        BytesStreamOutput out = new BytesStreamOutput();
+        diff.writeTo(out);
+        StreamInput in = StreamInput.wrap(out.bytes());
+        ImmutableOpenIntMap<TestDiffable> serialized = DiffableUtils.readImmutableOpenIntMapDiff(in, keySerializer,
+                new DiffableUtils.DiffableValueSerializer<Integer, TestDiffable>() {
+            @Override
+            public TestDiffable readFrom(StreamInput in, Integer key) throws IOException {
+                return new TestDiffable(in.readString());
+            }
+
+            @Override
+            public Diff<TestDiffable> readDiffFrom(StreamInput in, Integer key) throws IOException {
+                return AbstractDiffable.readDiffFrom(new StreamableReader<TestDiffable>() {
+                    @Override
+                    public TestDiffable readFrom(StreamInput in) throws IOException {
+                        return new TestDiffable(in.readString());
+                    }
+                }, in);
+            }
+        }).apply(before);
+        assertThat(serialized.size(), equalTo(3));
+        assertThat(serialized.get(1).value(), equalTo("1"));
+        assertThat(serialized.get(3).value(), equalTo("4"));
+        assertThat(serialized.get(4).value(), equalTo("5"));
+    }
+
+    public void testImmutableOpenIntMapDiffOfNonDiffables() throws IOException {
+        ImmutableOpenIntMap.Builder<String> builder = ImmutableOpenIntMap.builder();
```

Comment:
same.. we can maybe join the two tests into one method which we call with a parameter of whether to use diffs?